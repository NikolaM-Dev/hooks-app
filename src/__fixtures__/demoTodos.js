const demoTodos = [
  {
    id: 1,
    desc: 'Learn React',
    done: false,
  },
  {
    id: 2,
    desc: 'Learn Mongo',
    done: true,
  },
];

export default demoTodos;
