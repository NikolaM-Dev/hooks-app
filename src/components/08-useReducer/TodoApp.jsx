import { useEffect, useReducer } from 'react';

import AddTodo from './components/AddTodo';
import TodoList from './components/TodoList';
import todoReducer from './todoReducer';

import './TodoApp.css';

const init = () => {
  return JSON.parse(localStorage.getItem('todos')) || [];
};

const TodoApp = () => {
  const [todos, dispatch] = useReducer(todoReducer, [], init);

  useEffect(() => {
    localStorage.setItem('todos', JSON.stringify(todos));
  }, [todos]);

  const handleDelete = (todoId) => {
    dispatch({ type: 'delete', payload: todoId });
  };

  const handleToggle = (todoId) => {
    dispatch({ type: 'toggle', payload: todoId });
  };

  const handleAddTodo = (newTodo) => {
    dispatch({ type: 'add', payload: newTodo });
  };

  return (
    <>
      <h1>Todo App ({todos.length})</h1>
      <hr />

      <div className="row">
        <div className="col-7">
          <TodoList
            handleDelete={handleDelete}
            handleToggle={handleToggle}
            todos={todos}
          />
        </div>
        <div className="col">
          <AddTodo handleAddTodo={handleAddTodo} />
        </div>
      </div>
    </>
  );
};

export default TodoApp;
