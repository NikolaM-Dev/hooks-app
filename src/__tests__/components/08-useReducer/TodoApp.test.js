import '@testing-library/jest-dom';
import { mount, shallow } from 'enzyme';

import { act } from 'react-dom/test-utils';
import demoTodos from '__fixtures__/demoTodos';
import TodoApp from 'components/08-useReducer/TodoApp';

describe('Test en <TodoApp />', () => {
  Storage.prototype.setItem = jest.fn();

  let wrapper = shallow(<TodoApp />);

  beforeEach(() => {
    wrapper = shallow(<TodoApp />);
  });

  it('should show <TodoApp /> correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should add a to-do', () => {
    const wrapper = mount(<TodoApp />);

    act(() => {
      wrapper.find('AddTodo').prop('handleAddTodo')(demoTodos[0]);
      wrapper.find('AddTodo').prop('handleAddTodo')(demoTodos[1]);
    });

    const title = wrapper.find('h1').text().trim();

    expect(title).toBe('Todo App (2)');
    expect(localStorage.setItem).toHaveBeenCalledTimes(2);
  });

  it('should delete a to-do', () => {
    wrapper.find('AddTodo').prop('handleAddTodo')(demoTodos[0]);
    wrapper.find('TodoList').prop('handleDelete')(demoTodos[0].id);

    const title = wrapper.find('h1').text().trim();

    expect(title).toBe('Todo App');
  });
});
