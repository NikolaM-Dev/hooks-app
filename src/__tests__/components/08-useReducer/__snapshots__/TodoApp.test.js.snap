// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test en <TodoApp /> should show <TodoApp /> correctly 1`] = `
<Fragment>
  <h1>
    Todo App 
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col"
    >
      <AddTodo
        handleAddTodo={[Function]}
      />
    </div>
  </div>
</Fragment>
`;
